### PR TITLE
Adds more generic item-allocation queries

### PR DESCRIPTION
**Pull request by bnaecker**
- Creates the `omicron_nexus::db::queries` submodule, for custom /
  complex queries, organized around specific resources
- Adds a general-purpose `NextItem` query, which looks for the first
  available item in a database column, with a defined starting point and
  maximum offset / scan size. This can be used to implement a linear
  search over a defined space. One of the benefits is that for
  randomly-selected items, we can avoid a retry loop, by instead
  starting from the random base item and inserting the first successful
  value. There are several places where we use a similar query, or use a
  retry loop that could be replaced by this query.
- Use the `NextGuestIpv4Address` query when creating NICs.
- Use the `NextNicSlot` query when creating guest NICs

### CODE REVIEW

**Comment by bnaecker** [comment]
This PR is still pretty early, but I'd love to start getting some feedback. For context, this grew out of some work on #963. When we create MAC addresses for guests, we currently just fail the request if the address conflicts with an existing one. I started solving this the way we handle other such failures, with a retry loop. Even doing that, I came again across a familiar database query for picking the "next available item". That looks like [this](https://github.com/oxidecomputer/omicron/blob/ce3102ba5e6d0f33d361d94083d0d6ed44dd7fd9/nexus/src/db/subnet_allocation.rs#L752-L761).

I decided to bite the bullet and implement a more generic type for building these kinds of queries, which is [here](https://github.com/oxidecomputer/omicron/blob/8e59c301e92b2208b5d1e9a37203637dab4a9b53/nexus/src/db/queries/next_item.rs#L121). The idea is to describe the table and column we're searching, plus a starting point and maximum search size. I've implemented concrete versions of this currently for the IPv4 address and slot allocation for a guest network interface. If we're happy with where this goes, this can also be used to selection of:

- Guest NIC MAC addresses
- Guest NIC IPv6 addresses
- VPC Subnet IPv6 prefixes
- VPC VNIs
- VPC IPv6 /48 prefixes

I didn't do any of those here, since I'd like to get some feedback first. The key point is that I believe this allows us to avoid retrying the query in a loop in the application, and the attendant complexity and performance costs.

One point to note is related to #959. In that case, we have a specific single item that we'd like to _avoid_ in this allocation query. I have a good idea of how to implement that, by basically including a black-list of items and using something like a `WHERE <item> NOT IN <blacklist>` clause in the generated query. I'm not sure how well that performs, but for a handful of blacklist items it's probably acceptable.

Also note that a lot of the diff is reorganization. A number of bespoke queries were ending up in `omicron_nexus::db::subnet_allocation`, and I decided to split that out into resource / query-specific submodules under `omicron_nexus::db::queries`. To be concrete, `omicron_nexus::db::subnet_allocation` is now split into `omicron_nexus::db::queries::{vpc_subnet,network_interface}`. Those are unchanged, except for using the new `NextItem` query code internally.

File: nexus/src/db/queries/network_interface.rs [hunk 1]
```diff
@@ -642,6 +256,152 @@ fn decode_database_error(
...
+/// used to detect the case when there are no slots available.
+#[derive(Debug, Clone, Copy)]
+pub struct NextNicSlot {
+    inner: NextItem<
+        db::schema::network_interface::table,
+        i16,
+        db::schema::network_interface::dsl::slot,
+        Uuid,
+        db::schema::network_interface::dsl::instance_id,
+    >,
```
Comment (smklein) [inline nexus/src/db/queries/network_interface.rs:385]: Nice, this looks really compact for an allocation query!

File: nexus/src/db/queries/network_interface.rs [hunk 2]
```diff
@@ -642,6 +256,152 @@ fn decode_database_error(
...
+            inner: NextItem::new_scoped(0, instance_id, MAX_NICS_PER_INSTANCE),
+        }
+    }
+}
+
+impl QueryFragment<Pg> for NextNicSlot {
+    fn walk_ast<'a>(&'a self, mut out: AstPass<'_, 'a, Pg>) -> QueryResult<()> {
+        out.push_sql("SELECT COALESCE((");
+        self.inner.walk_ast(out.reborrow())?;
+        out.push_sql("), 0)");
```
Comment (smklein) [inline nexus/src/db/queries/network_interface.rs:400]: Isn't zero a valid slot? How would we distinguish allocation failure from "slot zero has been allocated"?

File: nexus/src/db/mod.rs [hunk 3]
```diff
@@ -18,10 +18,12 @@ pub mod ipv6;
 pub mod lookup;
 mod pagination;
 mod pool;
+// This is marked public because the error types are used elsewhere, e.g., in
+// sagas.
+pub mod queries;
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/db/mod.rs:23]: I think that could achieve what we need, yep.

File: nexus/src/db/queries/vni.rs [hunk 4]
```diff
@@ -0,0 +1,5 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Allocation of unique VPC Geneve Virtual Network Identifiers (VNI)
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/db/queries/vni.rs:5]: Ah, sorry placeholder. I started implementing these, but wanted to get eyes on before I dug further.

File: nexus/src/db/queries/vpc_subnet.rs [hunk 5]
```diff
@@ -0,0 +1,549 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/db/queries/vpc_subnet.rs:1]: That's correct. `vpc_subnet.rs` and `network_interface.rs` combine to what was `subnet_allocation.rs`.

File: nexus/src/db/queries/network_interface.rs [hunk 6]
```diff
@@ -642,6 +256,152 @@ fn decode_database_error(
...
+            // direct addition-based approach for IPv4 but not IPv6. This will
+            // loop, which, while it may not matter much, can be nearly
+            // trivially avoided by converting to u128, adding, and converting
+            // back. Given that these spaces can be _really_ big, that is
+            // probably worth doing.
+            network
+                .iter()
+                .nth(NUM_INITIAL_RESERVED_IP_ADDRESSES as _)
+                .unwrap_or_else(|| {
+                    panic!("Unexpectedly small IPv4 subnetwork: '{}'", network)
```
[the remark this answers is not in the document]
Comment (bnaecker) [reply nexus/src/db/queries/network_interface.rs:310]: Nice catch, thanks!

File: nexus/src/db/queries/network_interface.rs (hunk 1 above)
Comment (bnaecker) [reply nexus/src/db/queries/network_interface.rs:385]: Yeah, this is the meat of it. I'm hoping that we can just delegate almost everything to `NextItem`, with the right columns and search range types.

File: nexus/src/db/queries/network_interface.rs (hunk 2 above)
Comment (bnaecker) [reply nexus/src/db/queries/network_interface.rs:400]: Good question. Zero is indeed a valid slot. The `COALESCE` there is to handle the case where the `NextItem` query returns `NULL`. That only happens if there is no record in the table with the requested `instance_id`, i.e., there's not yet an interface for the instance (it's been verified to exist by the time we get here).

There is no "allocation failure," as I think you're imagining it. The `NextItem` query looks over the range `[0, 8]` for a slot. Note that it's inclusive on the upper bound, but that `8` is not a valid slot, `7` is the maximum allowed slot. So we'll return `8` if this instance is topped off in terms of interfaces, which will fail [this check](https://github.com/oxidecomputer/omicron/blob/a99069ac8967ea8aacc2724423e1b47fb30a9d7a/common/src/sql/dbinit.sql#L668) in the database. We detect that, and convert it to an error indicating that the instance does not support more interfaces.

Comment (smklein) [reply nexus/src/db/queries/network_interface.rs:400]: I think I'm understanding  - COALESCE catches the case where the instance doesn't exist, but the query was issued anyway.

If *that* happens, wouldn't this query return zero (risking colliding with the valid `[0, 7]` space of valid results)?

Comment (bnaecker) [reply nexus/src/db/queries/network_interface.rs:400]: It could be that I've messed up my logic, but here is my reasoning. `COALESCE(a, b)` returns `a` iff `a IS NOT NULL`, otherwise it returns `b`.

So, if the `NextItem` query returns `NULL`, we return `0`. The `NextItem` query looks like:

```sql
SELECT
    0 + offset AS slot
FROM
    generate_series(0, 8) AS slot
RIGHT OUTER JOIN
    network_interface
ON
    (instance_id, slot, time_deleted IS NULL) =
    (<instance_id>, 0 + offset, TRUE)
WHERE
    slot IS NULL
LIMIT 1
```

So, if there is an instance, then this returns the first value in the range `[0, 8]` where `network_interface.slot IS NULL`, since it's a right-outer join (return `NULL` for the left column where there is no match in the right part of the join-expression). For example, return `1` if there's an interface on the instance that's in slot `0`.

If the instance does not exist at all, then there are no rows in the subquery at all. That's because there are no records in `network_interface` with that `instance_id`. So the query is empty, i.e., it returns `NULL`. Thus the coalesce expression returns `0`.

In short, the _goal_ (again, I could have messed this up) is to return `0` if there is no interface at all for this instance (we're allocating the first), or `last_slot + 1` if we're allocating an _additional_ interface.

Comment (bnaecker) [reply nexus/src/db/queries/network_interface.rs:400]: I'm rereading your question. I think one part of the confusion is that the _instance_ does exist, it just doesn't have any interfaces. In that case, we _want_ to return zero here, since we're allocating the first.

Comment (bnaecker) [reply nexus/src/db/queries/network_interface.rs:400]: @smklein Let me know if this resolved your question. I'm planning to merge this, but we should make sure there's no other follow-up. Thanks!
